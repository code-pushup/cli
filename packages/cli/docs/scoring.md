# Scoring

**Scoring strategies** are mathematical approaches used to convert raw metrics into normalized scores
(typically 0-1 range) that provide meaningful feedback about performance against targets or budgets. These
strategies are particularly valuable in software development contexts where you need to track and visualize how
well various metrics (like bundle sizes, test coverage, API latency, or build times) are performing relative to
established thresholds or baselines. Each strategy offers different characteristics - some provide linear
relationships between the metric and score, while others offer more nuanced behaviors like soft caps or
logarithmic scaling.

## Value Scoring Strategies

### Overview Table

| **Strategy Name**     | **Formula**                                                                                                                            | **Description**                                                                |
| --------------------- | -------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------ |
| **Percent Used**      | $$\text{score}= \max\!\left(0,\,1 - \frac{S}{M}\right)$$                                                                               | Linearly falls from 1→0 as **S** goes 0→**M**, then clamps at 0.               |
| **Linear Overshoot**  | $$\text{score}= \begin{cases}1,&S\le M\\[4pt]1 - \dfrac{S - M}{M},&S>M\end{cases}$$                                                    | Holds 1 until **S**=**M**, then falls linearly to 0 by **S**=2·**M**.          |
| **Relative Baseline** | $$\text{score}= \mathrm{clamp}_{0..1}\!\Bigl(0.5 + \frac{\text{baseline} - S}{2\,\text{baseline}}\Bigr)$$                              | Centers at 0.5 when **S** matches baseline; >0.5 improvement, <0.5 regression. |
| **Sigmoid Soft-Cap**  | $$\text{score}= \frac{1}{1 + e^{\,k\,(S - M)}}$$                                                                                       | Remains near 1 before **M**, then gently decays, steep after threshold.        |
| **Logarithmic Decay** | $$\text{score}= \max\!\Bigl(0,\,1 - \frac{\log_{10}(1+S)}{\log_{10}(1 + M\,k)}\Bigr)$$                                                 | Penalises large **S** sharply, small overruns have mild impact.                |
| **Tiered Grading**    | $$\text{score}= \begin{cases}1,&P\ge0.9\\[2pt]0.75,&P\ge0.75\\[2pt]0.5,&P\ge0.5\\[2pt]0,&\text{otherwise}\end{cases},\;P=\frac{S}{M}$$ | Maps **S/M** into A/B/C/F bands for clear grade visualization.                 |

---

### Percent Used Details

Directly reflects remaining budget fraction.

**Pros:**

- Intuitive (% headroom).
- Simple to compute.  
  **Cons:**
- Flatlines at 0 once budget is exceeded.

**Implementation:**

```ts
function percentUsed(current: number, max: number): number {
  return Math.max(0, 1 - current / max);
}
```

**When to use it:**  
Ideal for any strict size or time budget (bundle size, build time) where you care about how much headroom remains. Simple dashboards and alerts benefit from a direct “% remaining” metric.

**Examples:**

- **Bundle Stats:** Track JS bundle headroom under a 500 KB limit.
- **Test Coverage:** % lines covered vs 80 % threshold.
- **Page Load (LCP):** Remaining ms under a 2 s budget.

**Chart:**

```mermaid
xychart-beta
    title "Percent Used (M=10)"
    x-axis [0,1,2,3,4,5,6,7,8,9,10]
    y-axis "Score" 0 --> 1
    line Percent [1,0.9,0.8,0.7,0.6,0.5,0.4,0.3,0.2,0.1,0]
```

---

### Linear Overshoot Details

Penalty grows linearly with overshoot.

**Pros:**

- Proportional and predictable.
- Easy to explain to stakeholders.  
  **Cons:**
- Can go negative if not clamped.

**Implementation:**

```ts
function linearOvershoot(current: number, max: number): number {
  if (current <= max) {
    return 1;
  }
  return Math.max(0, 1 - (current - max) / max);
}
```

**When to use it:**  
When you want to maintain proportional visibility into how far a metric has exceeded its limit (e.g., memory usage, API latency). Works well if you clamp at 0 to avoid negatives.

**Examples:**

- **API Latency:** Overshoot above 200 ms threshold.
- **Memory Usage:** RAM above configured max.
- **CSS Sizes:** Style sheets above a 50 KB cap.

**Chart:**

```mermaid
xychart-beta
    title "Linear Overshoot"
    x-axis [0,5,10,12,14,16,18,20]
    y-axis "Score" 0 --> 1
    line Overshoot [1,1,1,0.8,0.6,0.4,0.2,0]
```

---

### Relative Baseline Details

Compares current value to a historical baseline.

**Pros:**

- Symmetric: 0.5 = “no change.”
- Highlights improvements/regressions.  
  **Cons:**
- Requires reliable baseline data.

**Implementation:**

```ts
function relativeBaseline(current: number, baseline: number): number {
  const score = 0.5 + (baseline - current) / (2 * baseline);
  return Math.max(0, Math.min(1, score));
}
```

**When to use it:**  
Perfect for tracking trends across releases (bundle size, test performance, code complexity). Shifts focus from absolute budgets to continuous improvement.

**Examples:**

- **Bundle Stats:** Compare today’s bundle size vs last release.
- **Test Duration:** Measure speedup/regression of test suite.
- **Doc Coverage:** Track improvements in storybook coverage.

**Chart:**

```mermaid
xychart-beta
    title "Relative Baseline (baseline=10)"
    x-axis [0,1,2,3,4,5,6,7,8,9,10]
    y-axis "Score" 0 --> 1
    line Delta [1,0.95,0.9,0.85,0.8,0.75,0.7,0.65,0.6,0.55,0.5]
```

---

### Sigmoid Soft-Cap Details

Soft threshold with tunable steepness.

**Pros:**

- Gentle near the cap; harsh after.
- Never negative or >1.  
  **Cons:**
- Requires selecting a constant **k**.
- Less intuitive without plotting.

**Implementation:**

```ts
function sigmoidSoftCap(current: number, max: number, k: number = 0.5): number {
  return 1 / (1 + Math.exp(k * (current - max)));
}
```

**Parameter Details:**

- **k** (steepness factor): Controls how quickly the score drops after exceeding the threshold
  - `k = 0.1`: Very gentle decline, scores remain high even with significant overshoot
  - `k = 0.5`: Moderate decline (default), balanced between tolerance and penalty
  - `k = 1.0`: Steep decline, harsh penalty for any overshoot
  - `k = 2.0`: Very steep decline, almost binary behavior near the threshold

**When to use it:**  
Use when minor breaches are acceptable but you want strong pressure against large overruns (e.g. accessibility violations, deprecation warnings).

**Examples:**

- **ESLint Errors:** Small numbers of lint errors tolerated.
- **TS Diagnostics:** Minor type mismatches vs critical errors.
- **Security Audits:** Few low-severe alerts OK; many not.

**Chart:**

```mermaid
xychart-beta
    title "Sigmoid Soft-Cap (M=5, k=0.5)"
    x-axis [0,1,2,3,4,5,6,7,8,9,10]
    y-axis "Score" 0 --> 1
    line Sigmoid [0.92,0.88,0.82,0.73,0.62,0.50,0.38,0.27,0.18,0.12,0.08]
```

---

### Logarithmic Decay Details

Log-scale penalty for wide-range data.

**Pros:**

- Highlights massive overruns.
- Small ones barely noticed.  
  **Cons:**
- Formula less intuitive.

**Implementation:**

```ts
function logarithmicDecay(current: number, max: number, k: number = 2): number {
  const score = 1 - Math.log10(1 + current) / Math.log10(1 + max * k);
  return Math.max(0, score);
}
```

**Parameter Details:**

- **k** (scaling factor): Controls the logarithmic compression range
  - `k = 1`: Minimal compression, behaves more like linear decay
  - `k = 2`: Moderate compression (default), good balance for most metrics
  - `k = 5`: Strong compression, very tolerant of small overruns but penalizes large ones
  - `k = 10`: Maximum compression, almost flat until very large overruns

**When to use it:**  
Ideal for metrics spanning orders of magnitude (e.g. network RTT, bundle gzipped vs uncompressed sizes).

**Examples:**

- **CWV (FID):** First Input Delay distribution.
- **Bundle Stats:** Large vs small chunks on log scale.
- **API Throttling:** High‐tail latency events.

**Chart:**

```mermaid
xychart-beta
    title "Logarithmic Decay (M=5, k=2)"
    x-axis [0,1,2,3,4,5,6,7,8,9,10]
    y-axis "Score" 0 --> 1
    line Log [1,0.71,0.54,0.42,0.33,0.25,0.19,0.13,0.08,0.04,0]
```

---

### Tiered Grading Details

Discrete grade bands (A/B/C/F).

**Pros:**

- Very readable.
- No continuous math needed for viewers.  
  **Cons:**
- Abrupt jumps hide nuance.

**Implementation:**

```ts
function tieredGrading(current: number, max: number): number {
  const ratio = current / max;
  if (ratio >= 0.9) return 1;
  if (ratio >= 0.75) return 0.75;
  if (ratio >= 0.5) return 0.5;
  return 0;
}
```

**When to use it:**  
Best for executive dashboards or compliance checks where letter‐grade is preferred over percentages.

**Examples:**

- **Test Coverage:** A: ≥90 %, B: 75 – 89 %, C: 50 – 74 %.
- **Doc Coverage:** Storybook or JSDoc rates.
- **Security Score:** CVSS banding.

**Chart:**

```mermaid
xychart-beta
    title "Tiered Grading"
    x-axis [0,1,2,3,4,5,6,7,8,9,10]
    y-axis "Score" 0 --> 1
    line Grade [0,0,0,0,0.5,0.5,0.5,0.75,1,1,1]
```
