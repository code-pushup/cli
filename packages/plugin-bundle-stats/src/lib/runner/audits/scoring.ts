import type { Issue } from '@code-pushup/models';
import type { MinMax } from '../types.js';

export type BlacklistPatternList = readonly BlacklistEntry[];
export type BlacklistEntry =
  | string
  | {
      pattern: string;
      hint?: string;
    };

export type PenaltyConfig = {
  enabled?: boolean;
  artefactSize?: [number, number];
  warningWeight?: number;
  errorWeight?: number;
  /**
   * glob patterns when matching get penalised
   * e.g. packagenames outdates, should be lazy loaded, etc.
   * Can be simple strings or objects with pattern and optional hint
   */
  blacklist?: BlacklistPatternList;
};

export type ScoringConfig = {
  totalSize: MinMax;
  penalty?: false | PenaltyConfig;
};

export type ScoreCalculator = (value: number, issues: Issue[]) => number;

export const DEFAULT_PENALTY: PenaltyConfig = {
  warningWeight: 0.1,
  errorWeight: 0.2,
};

/**
 * Calculates normalized penalty score from issues using provided penalty options. Legacy function for backward compatibility.
 *
 * ## Penalty Parameters
 * - **E**: Count of issues of severity errors (🚨)
 * - **W**: Count of issues of severity warnings (⚠️)
 * - **we**: Weight per error issue (default 0.2)
 * - **ww**: Weight per warning issue (default 0.1)
 *
 * ## Default Penalty Options
 * When penalty options are not provided, these defaults are used:
 * - **errorWeight**: 0.2
 * - **warningWeight**: 0.1
 *
 * ## Issues Penalty Formula
 * ```
 * penalty = we × E + ww × W
 * normalizedPenalty = penalty / (we + ww)
 * ```
 *
 * @param issues - Array of diagnostic issues with severity levels
 * @param penalty - Penalty configuration with error and warning weights
 * @returns Normalized penalty score (0-1) based on issue counts and weights
 */
export function calculatePenalty(
  issues: Issue[],
  penalty?: PenaltyConfig,
): number {
  const errorWeight = penalty?.errorWeight ?? 0.2;
  const warningWeight = penalty?.warningWeight ?? 0.1;

  let errorCount = 0;
  let warningCount = 0;

  for (const issue of issues) {
    if (issue.severity === 'error') errorCount++;
    else if (issue.severity === 'warning') warningCount++;
  }

  const penaltyValue = errorWeight * errorCount + warningWeight * warningCount;
  const totalWeight = errorWeight + warningWeight;

  return totalWeight > 0 ? penaltyValue / totalWeight : 0;
}

/**
 * Creates a score calculator function configured with bundle settings. Applies direct penalty subtraction for intuitive scoring behavior with a maximum penalty cap of 20%.
 *
 * ## Scoring
 * Assigns a score in the range [0 … 1] to each artefact (or artefact selection) based on:
 * - Size vs. configurable minimum and maximum thresholds
 * - Direct penalty subtraction based on issue severity levels (when enabled)
 * - Penalty is capped at maximum 20% of the size score
 *
 * A perfect score (1) means "within acceptable range"; lower values indicate regressions.
 *
 * ## Size Parameters
 * - **S**: Actual bytes
 * - **Min**: Minimum threshold bytes (lower bound)
 * - **Max**: Maximum threshold bytes (upper bound)
 *
 * ## Size Score Formula
 * For single threshold (number):
 * ```
 * sizeScore = {
 *   1,  if S ≤ threshold
 *   0,  if S > threshold
 * }
 * ```
 *
 * For range thresholds [Min, Max]:
 * ```
 * sizeScore = {
 *   max(0, S/Min),  if S < Min (penalize under-sized bundles)
 *   1,              if Min ≤ S ≤ Max (perfect score)
 *   0,              if S > Max (too big = score 0)
 * }
 * ```
 *
 * ## Penalty Calculation
 * Direct subtraction approach with 20% maximum penalty cap:
 * ```
 * penaltyShift = errors × errorWeight + warnings × warningWeight
 * cappedPenalty = min(penaltyShift, sizeScore × 0.2)
 * ```
 * Default weights: errorWeight = 0.2, warningWeight = 0.1
 *
 * ## Final Score Calculation
 * ```
 * finalScore = max(0, sizeScore - cappedPenalty)
 * ```
 * This creates a penalty shift pattern where issues directly reduce the score by their weight values,
 * but the total penalty reduction is capped at 20% of the original size score.
 * Note: When `penalty` is `false`, only size score is used.
 *
 * @param options - Scoring configuration containing thresholds and penalty weights
 * @returns Score calculator function that takes (value, issues) and returns score [0-1]
 */
export function createBundleStatsScoring(
  options: ScoringConfig,
): ScoreCalculator {
  const { totalSize, penalty } = options;

  return (value: number, issues: Issue[] = []): number => {
    let sizeScore: number;

    if (Array.isArray(totalSize)) {
      // Range thresholds [min, max]
      const [minThreshold, maxThreshold] = totalSize;

      if (value < minThreshold) {
        sizeScore = Math.max(0, value / minThreshold);
      } else if (value <= maxThreshold) {
        sizeScore = 1.0;
      } else {
        sizeScore = 0;
      }
    } else {
      sizeScore = value <= totalSize ? 1.0 : 0;
    }

    if (penalty === false || issues.length === 0) {
      return sizeScore;
    }

    const penaltyOptions = penalty || DEFAULT_PENALTY;
    const penaltyShift = calculatePenaltyShift(issues, penaltyOptions);

    // Cap penalty at maximum 20% of the size score
    const maxPenalty = sizeScore * 0.2;
    const cappedPenalty = Math.min(penaltyShift, maxPenalty);

    return Math.max(0, sizeScore - cappedPenalty);
  };
}

/**
 * Calculates direct penalty shift based on issue counts and weights. Creates intuitive score reduction where each issue subtracts its weight from the final score.
 *
 * @param issues - Array of diagnostic issues with severity levels
 * @param options - Penalty configuration with error and warning weights
 * @returns Direct penalty shift value to subtract from size score
 */
function calculatePenaltyShift(
  issues: Issue[],
  options: PenaltyConfig,
): number {
  const errorCount = issues.filter(issue => issue.severity === 'error').length;
  const warningCount = issues.filter(
    issue => issue.severity === 'warning',
  ).length;

  const errorWeight = options.errorWeight ?? 0.2;
  const warningWeight = options.warningWeight ?? 0.1;

  // Direct penalty: each error/warning reduces score by its weight
  return errorCount * errorWeight + warningCount * warningWeight;
}
