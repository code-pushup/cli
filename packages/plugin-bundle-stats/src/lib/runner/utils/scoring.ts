import type { BundleStatsConfig } from '../types.js';

/**
 * Calculates the bundle score based on total size and threshold
 *
 * @param total - Total bundle size in bytes
 * @param config - Bundle configuration containing thresholds
 * @returns Score as a ratio (0-1) of actual size to threshold, capped at 1
 */
export function calculateScore(
  total: number,
  config: BundleStatsConfig,
): number {
  const threshold = config.thresholds.totalSize;
  let maxThreshold: number;

  if (Array.isArray(threshold)) {
    maxThreshold = threshold[1]; // Use max value from [min, max] tuple
  } else {
    maxThreshold = threshold;
  }

  // Score calculation: 1.0 if within threshold, decreasing as size increases
  if (total <= maxThreshold) {
    return 1.0;
  }

  // Calculate how much over the threshold we are
  const excess = total - maxThreshold;
  const excessRatio = excess / maxThreshold;

  // Score decreases linearly, but doesn't go below 0
  return Math.max(0, 1 - excessRatio);
}
