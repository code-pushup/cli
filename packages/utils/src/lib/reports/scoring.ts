import {
  AuditReport,
  CategoryConfig,
  CategoryRef,
  Group,
  GroupRef,
  PluginReport,
  Report,
} from '@code-pushup/models';
import { deepClone } from '../transform';

export type EnrichedAuditReport = AuditReport & { plugin: string };
export type WeighedAuditReport = EnrichedAuditReport & { weight: number };
export type EnrichedScoredGroupWithAudits = EnrichedScoredGroup & {
  audits: AuditReport[];
};
export type ScoredCategoryConfig = CategoryConfig & { score: number };

export type EnrichedScoredGroup = Group & {
  plugin: string;
  score: number;
};

export type ScoredReport = Omit<Report, 'plugins' | 'categories'> & {
  plugins: (Omit<PluginReport, 'audits' | 'groups'> & {
    audits: EnrichedAuditReport[];
    groups: EnrichedScoredGroup[];
  })[];
  categories: ScoredCategoryConfig[];
};

export function calculateScore<T extends { weight: number }>(
  refs: T[],
  scoreFn: (ref: T) => number,
): number {
  const { numerator, denominator } = refs.reduce(
    (acc, ref) => {
      const score = scoreFn(ref);
      return {
        numerator: acc.numerator + score * ref.weight,
        denominator: acc.denominator + ref.weight,
      };
    },
    { numerator: 0, denominator: 0 },
  );
  // No division by 0, otherwise we produce NaN
  // This can be caused by:
  // - empty category refs
  // - categories with refs only containing weight of `0`
  // both should get caught when validating the model
  if (!numerator && !denominator) {
    throw new Error(
      '0 division for score. This can be caused by refs only weighted with 0 or empty refs',
    );
  }
  return numerator / denominator;
}

export function scoreReport(report: Report): ScoredReport {
  const allScoredAuditsAndGroups = new Map();

  const scoredPlugins = report.plugins.map(plugin => {
    const { slug, audits, groups } = plugin;

    const updatedAudits = audits.map(audit => ({ ...audit, plugin: slug }));

    updatedAudits.forEach(audit => {
      allScoredAuditsAndGroups.set(`${slug}-${audit.slug}-audit`, audit);
    });

    function groupScoreFn(ref: GroupRef) {
      const score = allScoredAuditsAndGroups.get(
        `${slug}-${ref.slug}-audit`,
      )?.score;
      if (score == null) {
        throw new Error(
          `Group has invalid ref - audit with slug ${slug}-${ref.slug}-audit not found`,
        );
      }
      return score;
    }

    const scoredGroups =
      groups?.map(group => ({
        ...group,
        score: calculateScore(group.refs, groupScoreFn),
        plugin: slug,
      })) ?? [];

    scoredGroups.forEach(group => {
      allScoredAuditsAndGroups.set(`${slug}-${group.slug}-group`, group);
    });

    return { ...plugin, audits: updatedAudits, groups: scoredGroups };
  });

  function catScoreFn(ref: CategoryRef) {
    const key = `${ref.plugin}-${ref.slug}-${ref.type}`;
    const item = allScoredAuditsAndGroups.get(key);
    if (!item) {
      throw new Error(
        `Category has invalid ref - ${ref.type} with slug ${key} not found in ${ref.plugin} plugin`,
      );
    }
    return item.score;
  }

  const scoredCategories = report.categories.map(category => ({
    ...category,
    score: calculateScore(category.refs, catScoreFn),
  }));

  return {
    ...deepClone(report),
    plugins: scoredPlugins,
    categories: scoredCategories,
  };
}
