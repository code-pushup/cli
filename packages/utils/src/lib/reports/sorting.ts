import { AuditReport, CategoryRef, PluginReport } from '@code-pushup/models';
import {
  ScoredGroup,
  ScoredReport,
  SortableAuditReport,
  SortableGroup,
} from './types';
import {
  compareAudits,
  compareCategoryAuditsAndGroups,
  compareIssues,
  getSortableAuditByRef,
  getSortableGroupByRef,
} from './utils';

export function sortReport(report: ScoredReport): ScoredReport {
  const { categories, plugins } = report;
  const sortedCategories = categories.map(category => {
    const { audits, groups } = category.refs.reduce(
      (
        acc: {
          audits: SortableAuditReport[];
          groups: SortableGroup[];
        },
        ref: CategoryRef,
      ) => ({
        ...acc,
        ...(ref.type === 'group'
          ? {
              groups: [...acc.groups, getSortableGroupByRef(ref, plugins)],
            }
          : {
              audits: [...acc.audits, getSortableAuditByRef(ref, plugins)],
            }),
      }),
      { groups: [], audits: [] },
    );
    const sortedAuditsAndGroups = [...audits, ...groups].sort(
      compareCategoryAuditsAndGroups,
    );

    const sortedRefs = [...category.refs].sort((a, b) => {
      const aIndex = sortedAuditsAndGroups.findIndex(
        ref => ref.slug === a.slug && ref.plugin === a.plugin,
      );
      const bIndex = sortedAuditsAndGroups.findIndex(
        ref => ref.slug === b.slug && ref.plugin === b.plugin,
      );
      return aIndex - bIndex;
    });

    return { ...category, refs: sortedRefs };
  });

  return {
    ...report,
    categories: sortedCategories,
    plugins: sortPlugins(plugins),
  };
}

function sortPlugins(
  plugins: (Omit<PluginReport, 'audits' | 'groups'> & {
    audits: AuditReport[];
    groups: ScoredGroup[];
  })[],
) {
  return plugins.map(plugin => ({
    ...plugin,
    audits: [...plugin.audits].sort(compareAudits).map(audit =>
      audit.details?.issues
        ? {
            ...audit,
            details: {
              ...audit.details,
              issues: [...audit.details.issues].sort(compareIssues),
            },
          }
        : audit,
    ),
  }));
}
