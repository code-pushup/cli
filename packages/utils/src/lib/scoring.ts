import {
  AuditGroup,
  AuditGroupRef,
  AuditReport,
  CategoryConfig,
  CategoryRef,
  PluginReport,
  Report,
} from '@code-pushup/models';
import { deepClone } from './transformation';

export type EnrichedAuditReport = AuditReport & { plugin: string };
type ScoredCategoryConfig = CategoryConfig & { score: number };

export type EnrichedScoredAuditGroup = AuditGroup & {
  plugin: string;
  score: number;
};

export type ScoredReport = Omit<Report, 'plugins' | 'categories'> & {
  plugins: (Omit<PluginReport, 'audits' | 'groups'> & {
    audits: EnrichedAuditReport[];
    groups: EnrichedScoredAuditGroup[];
  })[];
  categories: ScoredCategoryConfig[];
};

export function calculateScore<T extends { weight: number }>(
  refs: T[],
  scoreFn: (ref: T) => number,
): number {
  const { numerator, denominator } = refs.reduce(
    (acc, ref) => {
      const score = scoreFn(ref);
      return {
        numerator: acc.numerator + score * ref.weight,
        denominator: acc.denominator + ref.weight,
      };
    },
    { numerator: 0, denominator: 0 },
  );
  return numerator / denominator;
}

export function scoreReport(report: Report): ScoredReport {
  const scoredReport = deepClone(report) as ScoredReport;
  const allScoredAuditsAndGroups = new Map();

  scoredReport.plugins?.forEach(plugin => {
    const { audits } = plugin;
    const groups = plugin.groups || [];

    audits.forEach(audit => {
      const key = `${plugin.slug}-${audit.slug}-audit`;
      audit.plugin = plugin.slug;
      allScoredAuditsAndGroups.set(key, audit);
    });

    function groupScoreFn(ref: AuditGroupRef) {
      const score = allScoredAuditsAndGroups.get(
        `${plugin.slug}-${ref.slug}-audit`,
      )?.score;
      if (score == null) {
        throw new Error(
          `Group has invalid ref - audit with slug ${plugin.slug}-${ref.slug}-audit not found`,
        );
      }
      return score;
    }

    groups.forEach(group => {
      const key = `${plugin.slug}-${group.slug}-group`;
      group.score = calculateScore(group.refs, groupScoreFn);
      group.plugin = plugin.slug;
      allScoredAuditsAndGroups.set(key, group);
    });
    plugin.groups = groups;
  });

  function catScoreFn(ref: CategoryRef) {
    const key = `${ref.plugin}-${ref.slug}-${ref.type}`;
    const item = allScoredAuditsAndGroups.get(key);
    if (!item) {
      throw new Error(
        `Category has invalid ref - ${ref.type} with slug ${key} not found in ${ref.plugin} plugin`,
      );
    }
    return item.score;
  }

  const scoredCategoriesMap = new Map();
  // eslint-disable-next-line functional/no-loop-statements
  for (const category of scoredReport.categories) {
    category.score = calculateScore(category.refs, catScoreFn);
    scoredCategoriesMap.set(category.slug, category);
  }

  scoredReport.categories = Array.from(scoredCategoriesMap.values());

  return scoredReport;
}
