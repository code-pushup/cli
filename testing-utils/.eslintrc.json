{
  "extends": ["../.eslintrc.json"],
  "ignorePatterns": ["!**/*"],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "parserOptions": {
        "project": ["testing-utils/tsconfig.*?.json"]
      }
    }
  ]
}
