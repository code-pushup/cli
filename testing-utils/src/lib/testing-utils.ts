export function testingUtils(): string {
  return 'testing-utils';
}
