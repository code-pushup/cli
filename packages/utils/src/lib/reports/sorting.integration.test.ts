import { REPORT_MOCK } from '@code-pushup/testing-utils';
import { scoreReport } from './scoring';
import { sortReport } from './sorting';

describe('sortReport', () => {
  it('should sort the audits and audit groups in categories, plugin audits and audit issues', () => {
    /*
      Sorting of the report is based on:
        - for category audits and audit groups: weight, score, value, title
        - for plugin audits: score, value, title
        - for audit issues: score, value, title

      The sorted report mock has the following data after sorting:

      Categories:

      Test results
      -> Cypress E2E: weight 3
      -> CyCT:        weight 1
      => score = 0.625

      Bug prevention
      -> TypeScript ESLint group:      weight 8
        -> ts-eslint-typing            weight 3
        -> ts-eslint-enums             weight 1
        -> ts-eslint-experimental      weight 0

      -> TypeScript ESLint functional: weight 1
      -> ESLint Jest naming:           weight 1
      -> ESLint Cypress:               weight 0
      => score = 0.3

      Plugins:

      Cypress results
      -> Cypress E2E: score 0.5, value 3
        -> error: Test `Display progress for selected commit` failed.
        -> error: Test `Sort audit table based on value` failed.
        -> error: Test `Open Bug prevention category detail` failed.
      -> CyCT:        score 1, value 0

      ESLint
      -> Type checking:                   score 0, value 2
        -> error: outputFile does not exist in type Cli.
        -> warning: command might be undefined.
      -> Functional principles:           score 0, value 1
        -> error: Unexpected let, use const instead.
      -> Typescript experimental checks   score 0, value 1
        -> info: Use better-enums.
      -> Consistent naming:               score 1, value 0
      -> Cypress rules:                   score 1, value 0
      -> Enumeration value checks:        score 1, value 0
    */
    const sortedReport = sortReport(scoreReport(REPORT_MOCK));
    expect(sortedReport).toMatchSnapshot();
  });

  it('should sort a report with no categories', () => {
    const sortedReport = sortReport(
      scoreReport({ ...REPORT_MOCK, categories: [] }),
    );
    expect(sortedReport).toEqual(expect.objectContaining({ categories: [] }));
  });
});
