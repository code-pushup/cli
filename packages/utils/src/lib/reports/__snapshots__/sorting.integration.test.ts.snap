// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sortReport > should sort the audits and audit groups in categories, plugin audits and audit issues 1`] = `
{
  "categories": [
    {
      "refs": [
        {
          "plugin": "cypress",
          "slug": "cypress-e2e-tests",
          "type": "audit",
          "weight": 3,
        },
        {
          "plugin": "cypress",
          "slug": "cypress-component-tests",
          "type": "audit",
          "weight": 1,
        },
      ],
      "score": 0.625,
      "slug": "test-results",
      "title": "Test results",
    },
    {
      "refs": [
        {
          "plugin": "eslint",
          "slug": "typescript-eslint",
          "type": "group",
          "weight": 8,
        },
        {
          "plugin": "eslint",
          "slug": "eslint-functional",
          "type": "audit",
          "weight": 1,
        },
        {
          "plugin": "eslint",
          "slug": "eslint-jest-consistent-naming",
          "type": "audit",
          "weight": 1,
        },
        {
          "plugin": "eslint",
          "slug": "typescript-eslint-extra",
          "type": "group",
          "weight": 0,
        },
        {
          "plugin": "eslint",
          "slug": "eslint-cypress",
          "type": "audit",
          "weight": 0,
        },
      ],
      "score": 0.3,
      "slug": "bug-prevention",
      "title": "Bug prevention",
    },
  ],
  "date": "2023-08-16T09:00:00.000Z",
  "duration": 666,
  "packageName": "@code-pushup/core",
  "plugins": [
    {
      "audits": [
        {
          "details": {
            "issues": [
              {
                "message": "Test \`Display progress for selected commit\` failed.",
                "severity": "error",
              },
              {
                "message": "Test \`Sort audit table based on value\` failed.",
                "severity": "error",
              },
              {
                "message": "Test \`Open Bug prevention category detail\` failed.",
                "severity": "error",
              },
            ],
          },
          "score": 0.5,
          "slug": "cypress-e2e-tests",
          "title": "Cypress e2e tests",
          "value": 3,
        },
        {
          "score": 1,
          "slug": "cypress-component-tests",
          "title": "Cypress component tests",
          "value": 0,
        },
      ],
      "date": "2023-08-16T09:00:00.000Z",
      "duration": 42,
      "groups": [],
      "icon": "cypress",
      "slug": "cypress",
      "title": "Cypress results",
    },
    {
      "audits": [
        {
          "details": {
            "issues": [
              {
                "message": "outputFile does not exist in type Cli",
                "severity": "error",
                "source": {
                  "file": "packages/cli/cli.ts",
                  "position": {
                    "endColumn": 10,
                    "endLine": 5,
                    "startColumn": 1,
                    "startLine": 1,
                  },
                },
              },
              {
                "message": "command might be undefined",
                "severity": "warning",
                "source": {
                  "file": "packages/cli/cli.ts",
                  "position": {
                    "endColumn": 20,
                    "endLine": 5,
                    "startColumn": 10,
                    "startLine": 5,
                  },
                },
              },
            ],
          },
          "score": 0,
          "slug": "typescript-eslint-typing",
          "title": "Type checking",
          "value": 2,
        },
        {
          "details": {
            "issues": [
              {
                "message": "Unexpected let, use const instead.",
                "severity": "error",
                "source": {
                  "file": "packages/core/report.ts",
                },
              },
            ],
          },
          "score": 0,
          "slug": "eslint-functional",
          "title": "Functional principles",
          "value": 1,
        },
        {
          "details": {
            "issues": [
              {
                "message": "Use better-enums.",
                "severity": "info",
              },
            ],
          },
          "score": 0,
          "slug": "typescript-eslint-experimental",
          "title": "TypeScript experimental checks",
          "value": 1,
        },
        {
          "score": 1,
          "slug": "eslint-jest-consistent-naming",
          "title": "Consistent naming",
          "value": 0,
        },
        {
          "score": 1,
          "slug": "eslint-cypress",
          "title": "Cypress rules",
          "value": 0,
        },
        {
          "score": 1,
          "slug": "typescript-eslint-enums",
          "title": "Enumeration value checks",
          "value": 0,
        },
      ],
      "date": "2023-08-16T09:00:00.000Z",
      "duration": 624,
      "groups": [
        {
          "refs": [
            {
              "slug": "typescript-eslint-typing",
              "weight": 3,
            },
            {
              "slug": "typescript-eslint-enums",
              "weight": 1,
            },
            {
              "slug": "typescript-eslint-experimental",
              "weight": 0,
            },
          ],
          "score": 0.25,
          "slug": "typescript-eslint",
          "title": "TypeScript ESLint",
        },
        {
          "refs": [
            {
              "slug": "typescript-eslint-experimental",
              "weight": 1,
            },
          ],
          "score": 0,
          "slug": "typescript-eslint-extra",
          "title": "TypeScript ESLint Extra",
        },
      ],
      "icon": "eslint",
      "slug": "eslint",
      "title": "ESLint",
    },
  ],
  "version": "1.0.0",
}
`;
